packaging/metadata.py

```diff
@@ -0,0 +1,387 @@
+# This file is dual licensed under the terms of the Apache License, Version
+# 2.0, and the BSD License. See the LICENSE file in the root of this repository
+# for complete details.
+
+import dataclasses
+import math
+import sys
+from email import message_from_bytes
+from email.headerregistry import Address, AddressHeader
+from email.message import EmailMessage
+from email.policy import EmailPolicy, Policy
+from functools import reduce
+from inspect import cleandoc
+from itertools import chain
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    ClassVar,
+    Collection,
+    Dict,
+    Iterable,
+    Iterator,
+    Mapping,
+    Set,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+    cast,
+)
+
+from .requirements import InvalidRequirement, Requirement
+from .specifiers import SpecifierSet
+from .version import Version
+
+T = TypeVar("T", bound="CoreMetadata")
+A = TypeVar("A")
+B = TypeVar("B")
+
+if sys.version_info[:2] >= (3, 8) and TYPE_CHECKING:  # pragma: no cover
+    from typing import Literal
+
+    NormalizedDynamicFields = Literal[
+        "platform",
+        "summary",
+        "description",
+        "keywords",
+        "home-page",
+        "author",
+        "author-email",
+        "license",
+        "supported-platform",
+        "download-url",
+        "classifier",
+        "maintainer",
+        "maintainer-email",
+        "requires-dist",
+        "requires-python",
+        "requires-external",
+        "project-url",
+        "provides-extra",
+        "provides-dist",
+        "obsoletes-dist",
+        "description-content-type",
+    ]
+else:
+    NormalizedDynamicFields = str
+
+
+def _normalize_field_name_for_dynamic(field: str) -> NormalizedDynamicFields:
+    """Normalize a metadata field name that is acceptable in `dynamic`.
+
+    The field name will be normalized to lower-case. JSON field names are
+    also acceptable and will be translated accordingly.
+
+    """
+    return cast(NormalizedDynamicFields, field.lower().replace("_", "-"))
+
+
+# Bypass frozen dataclass for __post_init__, this approach is documented in:
+# https://docs.python.org/3/library/dataclasses.html#frozen-instances
+_setattr = object.__setattr__
+
+
+# In the following we use `frozen` to prevent inconsistencies, specially with `dynamic`.
+# Comparison is disabled because currently `Requirement` objects are
+# unhashable/not-comparable.
+
+
+@dataclasses.dataclass(frozen=True, eq=False)
+class CoreMetadata:
+    """
+    Core metadata for Python packages, represented as an immutable
+    :obj:`dataclass <dataclasses>`.
+
+    Specification: https://packaging.python.org/en/latest/specifications/core-metadata/
+
+    Attribute names follow :pep:`PEP 566's JSON guidelines
+    <566#json-compatible-metadata>`.
+    """
+
+    # 1.0
+    name: str
+    version: Union[Version, None] = None
```

Comment:
The list of potential arguments is so huge it might be best to force keyword-only arguments for all but the project name.

```suggestion
    name: str
    _: KW_ONLY
    version: Union[Version, None] = None
```

Reply:
II had though about that, but I gave up because it seems that `KW_ONLY` is a very recent addition to stdlib...

Would this mean that packaging would not support Python <= 3.19?

Reply:
I'm fine with making the name kw-only as well. :)

Reply:
Argh. https://docs.python.org/3.7/library/dataclasses.html#dataclasses.field doesn't have any keyword stuff -- that came in 3.9/3.10. :/

Reply:
Yeap, unfoetunatelly Python docs don't use `.. versionadded:` for dataclasses constants, so it is a bit more difficult to track 😅

I also checked the source code to make sure they did not exist in previous versions.

Reply:
Darn, I thought I checked back to old versions.